### ISSUE TITLE
Target as hostname in configuration file

### ISSUE THREAD

**Issue by Rampa33** [rsyslog/rsyslog#3862]
When we provide hostname for target in rsyslog.conf, rsyslogd internally invokes getaddrinfo() to resolve the hostname to IP address. It is expected of DNS server to send a reply with multiple sockaddr. In that case, how is rsyslogd expected to work?

In code, i always see getaddrinfo() is invoked for first address in the response and we do not iterate the link list and try to connect to multiple destinations if present.

**Comment by davidelang** [comment]
That is correct, we will get one address (IIRC, the underlying call randomizes 
the order a bit, at least over time), and we use the first one that we get.

There is no attempt to connect to multiple addresses if they exist.

David Lang

**Comment by Rampa33** [comment]
@davidelang , thanks for your response. What if the first one in the list was unreachable at the moment of connection try? In that case, how will rsyslogd succeed in connecting to the remote device? 

Has it been done with the assumption that, in the next retry for the hostname, we could end up getting differently ordered list(first one in this order is luckily reachable).

**Comment by davidelang** [comment]
On Tue, 10 Sep 2019, Rampa33 wrote:

> @davidelang , thanks for your response. What if the first one in the list was unreachable at the moment of connection try? In that case, how will rsyslogd succeed in connecting to the remote device?
>
> Has it been done with the assumption that, in the next retry for the hostname, we could end up getting differently ordered list(first one in this order is luckily reachable).

Yes, that is the normal way that round-robin DNS is used.

I prefer other methods to provide high availability of the destination 
(CLUSTERIP if all the destinations are on the same subnet, or ANYCAST if they 
aren't) but RR DNS works within it's limits, and what you are describing is one 
of those limits.

David Lang
